File: detection/semgrep/fixtures/authz-scope-from-request-header.py
# Test fixture for rule: mcp-authz-scope-from-request-header-py
#
# S10 (forwarded-header-as-scope), Python/FastMCP spelling. FastMCP exposes the
# underlying HTTP request headers via get_http_headers() (a dict-like). A
# handler that trusts a tenant/identity header (X-Org-Id, or X-Forwarded-For
# for IP scoping) as the authorization scope has the same client-controlled
# input as the JS sibling. The legitimate-neighbor reads a non-scope header
# (X-Request-Id) for logging and must stay silent.

from fastmcp.server.dependencies import get_http_headers


# VULN: a scope-shaped header selects whose notes are returned.
def note_get_scoped_vuln(token):
    session = resolve_session(store, token)
    headers = get_http_headers()
    # ruleid: mcp-authz-scope-from-request-header-py
    header_org = headers.get("x-org-id")
    effective_org = header_org or session.org_id
    return store.list_notes_by_org(effective_org)


# OK: scope comes from the session; no header is consulted.
def note_get_scoped_fixed(token):
    session = resolve_session(store, token)
    # ok: mcp-authz-scope-from-request-header-py
    return store.list_notes_by_org(session.org_id)


# OK (legitimate neighbor): a non-scope header read for logging only.
def log_request():
    headers = get_http_headers()
    # ok: mcp-authz-scope-from-request-header-py
    req_id = headers.get("x-request-id")
    logger.info("tool invoked", extra={"req_id": req_id})

File: detection/semgrep/fixtures/authz-scope-from-request-header.js
// Test fixture for rule: mcp-authz-scope-from-request-header
//
// S10 (forwarded-header-as-scope). Over the MCP streamable-HTTP transport a
// tool handler receives the request's headers in extra.requestInfo.headers.
// A server behind a gateway is tempted to trust a tenant/identity header
// (X-Org-Id, or X-Forwarded-For in the IP-scoping variant) as the
// authorization scope — but that header is fully client-controlled. The
// legitimate-neighbor case reads a NON-scope header (X-Request-Id) for
// logging, which must stay silent: the rule keys on the header NAME, not on
// the mere act of reading a header.

// VULN: a scope-shaped header selects whose notes are returned.
async function noteGetScopedVuln({ token }, extra) {
  const session = resolveSession(store, token);
  // ruleid: mcp-authz-scope-from-request-header
  const headerOrg = extra?.requestInfo?.headers?.["x-org-id"];
  const effectiveOrgId = headerOrg ? headerOrg : session.orgId;
  return store.listNotesByOrg(effectiveOrgId);
}

// VULN (IP-scoping variant): X-Forwarded-For trusted for the same purpose.
async function noteGetScopedByIpVuln({ token }, extra) {
  const session = resolveSession(store, token);
  // ruleid: mcp-authz-scope-from-request-header
  const fwd = extra.requestInfo.headers["x-forwarded-for"];
  return store.listNotesForClientIp(fwd ?? session.clientIp);
}

// OK: scope comes from the session; no header is consulted at all.
async function noteGetScopedFixed({ token }, extra) {
  const session = resolveSession(store, token);
  // ok: mcp-authz-scope-from-request-header
  return store.listNotesByOrg(session.orgId);
}

// OK (legitimate neighbor): a non-scope header read for logging only. The
// scope-header name regex must not match "x-request-id".
async function logRequest(_args, extra) {
  // ok: mcp-authz-scope-from-request-header
  const reqId = extra?.requestInfo?.headers?.["x-request-id"];
  logger.info({ reqId, msg: "tool invoked" });
}
